### ISSUE TITLE
Potential bug with regex in nav Walker?

### ISSUE THREAD

**Issue by Sammaye** [understrap/understrap#884]
Please delete anything that does not apply for you!

## Issue is about..
  - [x ] a Bug Report
  - [ ] a Feature Request
  - [ ] a Suggestion
  - [ ] a Question
  - [ ] I need help

## Please tell us about your environment:

  - **WordPress Version:** 5.0.1
  - **OS Version:** Linux [Ubunrtu/16.04LTS]
  - **Browser:** all
  - **Language:** PHP
  - **Node.js & NPM Versions:** N/A

---

# Issue Overview for Bugreports

## Steps to reproduce?

Have a menu with two or more sub menus

## *Who* is the bug affecting?

Anyone with this most likely

## *What* is affected by this bug?

The assigning of the ID will be the same as the first sub menu at all times.

## *When* does this occur?

Whenever there are more than one sub menus

## *Where* on the platform does it happen?

dunno

## *How* do we replicate the issue?

Have a menu with two or more sub menus

## Are there any related Issues and/or PRs?

Not from what I can see

## Expected behavior

It should get the ID of the element in question

### Other Comments

Nice extension, I actually did my own and then thought of seeing if someone already did this haha, anyway I was looking at what you did in https://github.com/understrap/understrap/blob/master/inc/class-wp-bootstrap-navwalker.php#L77 and I noticed that that regex might not work the way you expect since `$output` is actually the nav HTML so far, which means that if you have multiple sub menus it will always pick the first one.

TBH, I better solved this by just storing the current $item in a class variable, solves the whole guessing game and slow regexes, and takes up an extra like 0.1kb of memory which can be easily gc'ed by PHP when it's done.

**Comment by Sammaye** [comment]
Looking at it more I noticed it might break for menus with deeper levels since it doesn't skip them, most browsers, like Chrome will fix the HTML but you must remember that Google will hate you

**Comment by Sammaye** [comment]
Also the coding standards of this class are very questionable, like https://github.com/understrap/understrap/blob/master/inc/class-wp-bootstrap-navwalker.php#L222

**Comment by pattonwebz** [comment]
Hey, if you could post these on the source repo for the walker I'll be better able to keep track of these and handle them when I have time to update the walker. I'm likely to miss them when I make updates to the walker if they exist only in this repo.

Thanks in advance :)

**Comment by Sammaye** [comment]
Ok kool, I'll sum them up and put them there thanks

**Comment by Thomas-A-Reinert** [comment]
Thanks for your input @Sammaye !
But it would also be pretty kind if you could sum up all your input and just leave one or two posts.. and not one every other minute ;)
Really, we´re not perfect and Understrap is by far not flawless, but it would be cool to _not_ receive email updates concerning _one_ issue in such short intervals .. No offense intended and I really hope you understand what I mean. We´re listening and try to improve anyway.

**Comment by Sammaye** [comment]
Yeah, sorry about that, sometimes I can run off, I saw one thing and then decided to look more and it just kept coming, I will sum it all up

**Comment by Thomas-A-Reinert** [comment]
Hey, everything cool - I totally understand and we really appreciate it when someone finds stuff we may have missed. As a sidenote: pull requests would also be appreciated if you find ways to improve our repo (or @pattonwebz in this case) ;)

**Comment by Sammaye** [comment]
I might make a PR though, I have only just started on WP really, I made my first WP site in a hurry for a client and now I am actually learning it for real, I am finding it very hard to produce clean and nice looking code in WP haha, put I might put up what I have atm

**Comment by pattonwebz** [comment]
@Sammaye You are not the first person to be surprised by the way WordPress PHP code is written and how it functions. It is not exactly like the other PHP code you work with in other projects - it's very procedural, kind of old style (being that WP still supports as far back as 5.4 - to the distain of many a developer lol) and coding standards are somewhat obscure from PSRs. It's the world we live in though and we deal with it. Recently a lot of developers have started to look at PSR4 support but other than that one we mostly do the opposite in WP haha!

In general WP is a great community to work in though because of how big into opensource it is. Issue reports are always welcome as well as code contributions. We aren't always able to act on them immediately but in almost all cases the are still appreciated (and personally I like PRs that fix my spelling mistakes. I make spelling mistakes way too often lol ).

**Comment by Thomas-A-Reinert** [comment]
@Sammaye - you´re welcome, really. In case you need any help - just holla. You may want to take a peep at the "[upcoming documentation](https://usdocu.tarthemes.com/)" (Put in quotes, because we are _really_ off _any_ schedule where we _originally_ wanted to release it.). It´s still not complete and it will potentially have a load of asynchronisms off the actual state of the project. But hey - sometimes you got some real-life-things to fix before anything else. And my personal professional interest moved to anything but WP, so i´m kinda short of time to spend into WP.

In case you find any flaws, things missing, things to add, spelling mistakes ;) or whatsoever in the docs - you´re welcome to report them [to the original repo](https://github.com/Thomas-A-Reinert/Understrap-Documentation) . Be it a pull request or just an issue. I´ll fix that asap.  Promised :)

My 2cents about the "backwards-compatibility problem".. 
It´s a blessing as well as it´s a real **problem**. 

**The bad thing:**
No other CMS has such low-level requirements as WP has. Be it Drupal, Typo3 or _any_ other. That makes the usual John Doe set up a WP site with a one-click install with his host and let it install plugins and themes galore - and without even knowing what s/he´s doing. And let it prolly abandon it. Which will probably be hacked sooner or later. Which again, will account to the negative stats of WP..

\<rant-start>Or even be able to judge the quality of the installed thingies. And yes, I can BUY a gazillion 5 Star ratings for a few bucks. Things are sometimes TOO easy with WP.

The lowest level of those would be called "DAU´s" in german which would prolly best translate into english with "dumbest assumbale user" or "PEBKAC". That´s usually - to my experience food bloggers. They don´t have _any_ concept upfront. They just want to jumpstart their thing and turn it - content-wise as well as visually - into any possible direction. Which makes themes like DIVI and the likes so extremely popular.\</rant-end>

**The positive**
My customers usually get a grip of how to write an article/page even with SEO-optimization within an hour. And they can even recall that when they haven´t been working with the site for more than two weeks. Try to achieve that with T3 or Drupal. Users will get lost there again and again and need a new training course here and there.

**The real negative**
\<rant-start>And that´s also the reason why Wp-Devs don´t get paid to what they deserve in Germany.
Freelance Typo3/Drupal Devs make about 800€/day while it´s hard to even get a 400€/day for any WP job.

And it doesn´t even matter if you can provide evidence for your profession. You´ll compete against any other WP-Hobbyist. \</rant-end>
